### PR TITLE
show politeia proposals

### PR DESCRIPTION
**Pull request by JustinBeBoy**
Requires planetdecred/dcrlibwallet#161

**This PR allows users to:**

- Add Proposal politeia in setting
- View list proposal politeia
- View detail proposal politeia
- Set up receiving notifications (new proposal, starting and ending votes)
- Share and open proposal politeia in browser

**How the notifications work:**

When the client launches the app for the first time, a sync is initiated and all proposals are fetched, now when next the client launches the app the sync initiates again, and if there are new proposals or changes to the proposals fetched initially, our callback listeners are triggered

We have listeners for onNewProposal, onVoteStarted, and onVoteEnded . When any of the listeners are triggered, it sends the appropriate notification to the client

When the client taps on a particular notification, it takes them to the proposal details page for that particular proposal

**Screenshots**

### CODE REVIEW

**Review by macsleven — CHANGES_REQUESTED** [review]

File: Decred Wallet/Features/Navigation Menu/NavigationMenuTabBarController.swift [hunk 1]
```diff
@@ -139,6 +139,9 @@ class NavigationMenuTabBarController: UITabBarController {
         // start sync and new tx listener
         SyncManager.shared.startSync(allowSyncOnCellular: Settings.syncOnCellular)
         TransactionNotification.shared.startListeningForNotifications()
+        
+        // start sync politeia
+        PoliteiaNotification.shared.syncPoliteia()
```
Comment (macsleven) [inline Decred Wallet/Features/Navigation Menu/NavigationMenuTabBarController.swift:144]: this should be done online when the wallet is online

File: Decred Wallet/Wallet Utils/PoliteiaNotification.swift [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+class PoliteiaNotification: NSObject {
+    static let shared = PoliteiaNotification()
+    
+    func syncPoliteia() {
+        DispatchQueue.global(qos: .userInitiated).async {
+            do {
+                try WalletLoader.shared.multiWallet.politeia?.sync(self)
+            } catch let error {
+                print("PoliteiaNotification sync Error:", error.localizedDescription)
```
Comment (macsleven) [inline Decred Wallet/Wallet Utils/PoliteiaNotification.swift:20]: This error should be handled properly, display the error message to the user using the banner, there should be a retry button somewhere in the app also.

**Review by macsleven — CHANGES_REQUESTED** [review]

File: Decred Wallet/Wallet Utils/PoliteiaNotification.swift (hunk 2 above)
Comment (JustinBeBoy) [reply Decred Wallet/Wallet Utils/PoliteiaNotification.swift:20]: I think it should only display the message: "There was an error when sync politeia and politeia would be synced the next time"

File: Decred Wallet/Wallet Utils/PoliteiaNotification.swift [hunk 3]
```diff
@@ -12,21 +12,33 @@ import Dcrlibwallet
...
             do {
                 try WalletLoader.shared.multiWallet.politeia?.sync()
+                self.setSyncStatus(isSyncSuccess: true)
             } catch let error {
                 print("PoliteiaNotification sync Error:", error.localizedDescription)
+                self.setSyncStatus(isSyncSuccess: false)
                 DispatchQueue.main.async {
                     if let navigationTabController = NavigationMenuTabBarController.instance?.view {
-                        Utils.showBanner(in: navigationTabController, type: .error, text: "There was an error when sync politeia and politeia would be synced the next time")
+                        Utils.showBanner(in: navigationTabController, type: .error, text: "There was an error when sync politeia, please try again in Politeia")
```
Comment (macsleven) [inline Decred Wallet/Wallet Utils/PoliteiaNotification.swift:25]: localize string

File: Decred Wallet/Wallet Utils/PoliteiaNotification.swift [hunk 4]
```diff
@@ -12,21 +12,33 @@ import Dcrlibwallet
 class PoliteiaNotification: NSObject {
     static let shared = PoliteiaNotification()
     
+    private let IS_SYNC_SUCCESS_KEY = "sync_politeia_success"
```
Comment (macsleven) [inline Decred Wallet/Wallet Utils/PoliteiaNotification.swift:15]: this key should be gotten from dcrilibwallet

**Review by macsleven — CHANGES_REQUESTED** [review]

File: Decred Wallet/Extensions/Dcrlibwallet.swift [hunk 5]
```diff
@@ -173,6 +173,54 @@ extension DcrlibwalletWallet {
     }
 }
 
+extension DcrlibwalletPoliteia {
+    func getPoliteias(category: PoliteiaCategory = PoliteiaCategory.pre, offset: Int32 = 0, limit: Int32 = 20, newestFirst: Bool = true) -> Array<Politeia>? {
+        var error: NSError?
+        let politeias = self.getProposals(category.rawValue, offset: offset, limit: limit, newestFirst: newestFirst, error: &error)
+        if error != nil {
+            print("getProposals error: ", error!.localizedDescription)
```
Comment (macsleven) [inline Decred Wallet/Extensions/Dcrlibwallet.swift:181]: display this to user

File: Decred Wallet/Extensions/Dcrlibwallet.swift [hunk 6]
```diff
@@ -173,6 +173,54 @@ extension DcrlibwalletWallet {
...
+        
+        if politeias.isEmpty {
+            return []
+        }
+
+        do {
+            let response = try JSONDecoder().decode(Array<Politeia>.self, from: politeias.data(using: .utf8)!)
+            return response
+        } catch let error {
+            print("decode allPoliteia error:", error.localizedDescription)
```
Comment (macsleven) [inline Decred Wallet/Extensions/Dcrlibwallet.swift:193]: display this to user

File: Decred Wallet/Extensions/Dcrlibwallet.swift [hunk 7]
```diff
@@ -173,6 +173,54 @@ extension DcrlibwalletWallet {
...
+    func categoryCount(category: PoliteiaCategory) -> String {
+        let int32Pointer = UnsafeMutablePointer<Int32>.allocate(capacity: 4)
+        defer {
+            int32Pointer.deallocate()
+        }
+        
+        do {
+            try self.count(category.rawValue, ret0_: int32Pointer)
+        } catch let error {
+            print("count tx error:", error.localizedDescription)
```
Comment (macsleven) [inline Decred Wallet/Extensions/Dcrlibwallet.swift:207]: display this to user

File: Decred Wallet/Extensions/Dcrlibwallet.swift [hunk 8]
```diff
@@ -173,6 +173,54 @@ extension DcrlibwalletWallet {
...
+        return "\(category.description) (\(int32Pointer.pointee))"
+    }
+    
+    func detailPoliteia(_ id: Int) -> Politeia? {
+        do {
+            let politeia = try self.getProposalByIDRaw(id)
+            
+            return Politeia(politeia)
+        } catch let error {
+            print("decode allPoliteia error:", error.localizedDescription)
```
Comment (macsleven) [inline Decred Wallet/Extensions/Dcrlibwallet.swift:218]: display this to user

File: Decred Wallet/Extensions/Dcrlibwallet.swift [hunk 9]
```diff
@@ -173,6 +173,54 @@ extension DcrlibwalletWallet {
...
+        } catch let error {
+            print("count tx error:", error.localizedDescription)
+        }
+        return "\(category.description) (\(int32Pointer.pointee))"
+    }
+    
+    func detailPoliteia(_ id: Int) -> Politeia? {
+        do {
+            let politeia = try self.getProposalByIDRaw(id)
+
```
Comment (macsleven) [inline Decred Wallet/Extensions/Dcrlibwallet.swift:215]: extra white space not needed

File: Decred Wallet/Features/More/More Features/Politeia/Politeia.storyboard [hunk 10]
```diff
@@ -0,0 +1,540 @@
...
+
+
+Lorem ipsum dolor sit er elit lamet, consectetaur cillium adipisicing pecu, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. Nam liber te conscient to factor tum poen legum odioque civiuda.
+
+Lorem ipsum dolor sit er elit lamet, consectetaur cillium adipisicing pecu, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. Nam liber te conscient to factor tum poen legum odioque civiuda.</string>
```
Comment (macsleven) [inline Decred Wallet/Features/More/More Features/Politeia/Politeia.storyboard:464]: this is not needed

File: Decred Wallet/Features/More/More Features/Politeia/PoliteiaController.swift [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+        self.updateSyncingStatus()
+    }
+    
+    func startListeningForNotifications() {
+        try? WalletLoader.shared.multiWallet.politeia?.add(self, uniqueIdentifier: "\(self)")
+    }
+    
+    func setupArrayFilter() {
+        guard let politeia = WalletLoader.shared.multiWallet.politeia else {
+            print("PoliteiaController.setupFilter get politeia instance false")
```
Comment (macsleven) [inline Decred Wallet/Features/More/More Features/Politeia/PoliteiaController.swift:67]: log this error to dcrlibwallet, so we can get it from the user if it occurs

File: Decred Wallet/Features/More/More Features/Politeia/PoliteiaController.swift [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+        let barButtonTitle = UIBarButtonItem(title: LocalizedStrings.politeia, style: .plain, target: self, action: nil)
+        barButtonTitle.tintColor = UIColor.appColors.darkBlue
+        
+        self.navigationItem.leftBarButtonItems =  [closeButton, barButtonTitle]
+    }
+    
+    override func viewWillDisappear(_ animated: Bool) {
+        super.viewWillDisappear(animated)
+        WalletLoader.shared.multiWallet.politeia?.removeNotificationListener("\(self)")
+
```
Comment (macsleven) [inline Decred Wallet/Features/More/More Features/Politeia/PoliteiaController.swift:162]: extra white space not needed

File: Decred Wallet/pl.lproj/Localizable.strings [hunk 13]
```diff
@@ -395,6 +411,7 @@
 "seeDocumentations" = "Wyświetl dokumentacje";
 "beepForNewBlocks" = "Powiadamiaj dźwiękiem o nadejściu nowych bloków";
 "checkStatistics" = "Wyświetl statystyki";
+"enablePoliteiaNotification" = "Enable politeia notifications";
```
Comment (macsleven) [inline Decred Wallet/pl.lproj/Localizable.strings:414]:
```suggestion
"enablePoliteiaNotification" = "Enable politeia notifications";
"checkStatistics" = "Wyświetl statystyki";
```

**Review by macsleven — CHANGES_REQUESTED** [review]

**Review by dreacot — CHANGES_REQUESTED** [review]
Only Approved and Rejected Proposals should have the vote bar. Both in the proposal list page and the proposal details page

**Review by dreacot — CHANGES_REQUESTED** [review]
On the proposal list page, modify the info in the votebar to look like this
Screenshot from 2021-02-26 21-18-12

**Review by dreacot — CHANGES_REQUESTED** [review]
I can see a proposal under the voting category with a label of approved, please check it out, to ensure syncing is happening properly

Screenshot from 2021-02-26 21-32-34

**Review by dreacot — CHANGES_REQUESTED** [review]
Add some padding to the drop down menu, take a look at the transactions page for reference

Screenshot from 2021-02-26 21-38-50

**Review by dreacot — CHANGES_REQUESTED** [review]
On the proposal details page, remove the text in the center of the vote bar, it has been removed as well on android

**Comment by dreacot** [comment]
> Only Approved and Rejected Proposals should have the vote bar. Both in the proposal list page and the proposal details page
> 
> 

This issue still seems to be there!
steps to replicate
1. go to the politeia page
2. switch to approved category
3. then switch to abandoned category, the vote bar still shows

**Comment by dreacot** [comment]
> I can see a proposal under the voting category with a label of approved, please check it out, to ensure syncing is happening properly
> 
> Screenshot from 2021-02-26 21-32-34

This problem still seems to be occurring, see screenshot below

Screenshot from 2021-03-01 15-16-12

If you are certain the issue isn't from the ios code, then it might be work bringing up in the dcrlibwallet room so it can be properly looked at

**Comment by dreacot** [comment]
> > I can see a proposal under the voting category with a label of approved, please check it out, to ensure syncing is happening properly
> > Screenshot from 2021-02-26 21-32-34
> 
> This problem still seems to be occurring, see screenshot below
> 
> Screenshot from 2021-03-01 15-16-12
> 
> If you are certain the issue isn't from the ios code, then it might be worth bringing up in the dcrlibwallet room so it can be properly looked at

@JustinBeBoy This still hasn't been addressed

**Comment by dreacot** [comment]
When there's no proposal for a category the message should say "No proposals for selected category"

**Review by dreacot — APPROVED** [review]

**Review by macsleven — APPROVED** [review]

**Pull request merged**